src/crimson/common/gated.h

```diff
@@ -52,4 +68,43 @@ class Gated {
   seastar::gate pending_dispatch;
 };
 
-}// namespace crimson::common
+class gate_per_shard : public Gated {
+ public:
+  gate_per_shard() : gates(seastar::smp::count) {}
+  explicit gate_per_shard(size_t shard_count) : gates(shard_count) {}
+  gate_per_shard(const gate_per_shard&) = delete;
+  gate_per_shard& operator=(const gate_per_shard&) = delete;
+  gate_per_shard(gate_per_shard&&) = default;
+  gate_per_shard& operator=(gate_per_shard&&) = default;
+  ~gate_per_shard() override = default;
+
+  template <typename Func, typename T>
+  inline auto dispatch_in_background(const char* what, T& who, Func&& func) {
+    return dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func, typename T>
+  inline auto dispatch(const char* what, T& who, Func&& func) {
+    return gates[seastar::this_shard_id()].dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func>
+  auto submit_with_gate(const char* what, Func&& func) {
```

Comment:
Nit, follow the name convention above using "dispatch".

Probably `simple_dispatch(what, func)` because it doesn't deal with `system_shutdown_exception` and `who`?

Reply:
sure, sounds good